### ISSUE TITLE
Are newtypes always an option?

### ISSUE THREAD

**Issue by Ixrec** [Ixrec/rust-orphan-rules#5]
As far as I know, any time you run into an orphan rules violation, it is always _possible_ to solve it by writing a newtype. It might be [tedious to write](https://github.com/Ixrec/rust-orphan-rules/issues/3), or [not the best possible solution], but I'm not aware of a case where that's _impossible_ to solve with a newtype.

This issue is for proving me wrong.

**Comment by FrankHB** [comment]
NO.

I don't want to detail too mach, but there should be the simple case: for a type `W<A>` where `W` is a library generic type depending on client (user-provided) type `A` for function right, it may be not appropriate to introduce any new types for either `W`, `A`, or `W<A>` because the remaining program relies on their *type identity*.

For example, sane language binding implementation distinguish strictly about any "newtype" and wrapped type as different ones, for clear logical reasons. `A` is not new type for `A`. `W<A>` is not new type for `W<A>`. This should be true in any nominal type systems, including Rust; and an object language will assume this, too. Then, where should `W<A>` be settled down?

New types cannot be introduced to resolve the problem because:

* For `W`: Introducing new type implies that the library (crate) defining `W` cannot take the signature including `W` for a stable interface boundary. In some cases, concrete type `W` can be generalized to be a parameterized type so it won't be referenced directly in the library code whose signature containing `W`, but given the fact that Rust is lacking of structural typing in generics, it must come with some additional traits, say, `T` for one instance. Then, such trait plays the same role as `W` and suffers from the problem of `W<A>` (see below): what if `T<A>` is to be implemented? *Where?*
* For `A`: Common sense: the library author cannot put arbitrary `A` into the library. Neither the newtype for `A`.
* For `W<A>`: The library author also have nothing to do with `W<A>`. There is an extensibility issue for client code: It *should* be together with `A`. But it is not allowed. Why on earth `W<A>` does not qualified to be a new type for the sake of providing the implementation? Nope. It just be clashed with the orphan rule, occasionally.

**Comment by rft-sbohne** [comment]
Take a look at the [ordered-float](https://github.com/reem/rust-ordered-float) crate. They are one of the few crates that do an excellent job at providing trait implementations for 16 3rd-party crates. But the source code is almost 3000 lines of boilerplate code around what is essentially these 4 lines (comments omitted)
```
    #[inline]
    fn ge(&self, other: &Self) -> bool {
        self.0.is_nan() | (self.0 >= other.0)
    }
```

And it's not enough. If I wanted to use `OrderedFloat` with [utoipa](https://github.com/juhaku/utoipa) then I have a problem, because utoipa uses their own [ToSchema](https://docs.rs/utoipa/latest/utoipa/trait.ToSchema.html) trait which is not supported by `ordered-float`, instead of [schemars](https://docs.rs/schemars/latest/schemars/)'s trait which is supported. I can't see another layer of newtype as a viable and maintable solution given that the first layer already took 3000 lines of code.

In short: newtypes don't scale with the number traits (which is ever growing thanks to Rust's great community).

The story would be different if most traits would support a derive for newtypes. Those 3000 lines of boilerplate could and should be a couple lines of `#[derive(...)]`. But not even the standard library does that (in particular the std::ops traits and Display should support derive for newtypes). And if the standard library doesn't support the newtype pattern, why would 3rd-party libraries support it?
